[Justin Peng] iP

JustinPeng13 opened the pull request:
# MakiBot🤖💬
### Features
- **Add** Todos, Deadlines, & Events
- **Mark** tasks as done/undone
- **Save/list/delete** tasks
- **Find** tasks
### Try It Out
1. Download the latest [release](https://github.com/JustinPeng13/iP/releases/tag/A-Jar) from the [repo](https://github.com/JustinPeng13/iP)
2. Move `duke.jar` to a folder of your choice
3. In the same folder, run `java -jar duke.java`
4. Manage your tasks with MakiBot! 
#### Output
> Hello! I'm MakiBot
You are currently in timezone: GMT
Would you like to change your timezone? Y/N
#### Main Function
```
public static void main(String[] args) {
    new Duke().start();
}
```
#### What are you waiting for?
- [x] I'll try MakiBot out now!

Commit:
Add Gradle support

JustinPeng13 committed:
Added start method to greet, echo, and handle exit commands

JustinPeng13 committed:
Added taskList to store tasks and list function

JustinPeng13 committed:
Created Task class, allow marking as done and undone

JustinPeng13 committed:
Added Todo, Deadline, and Event classes

JustinPeng13 committed:
Set up test case for runtest.bat

JustinPeng13 committed:
Create exceptions + handling, abstracted commands

JustinPeng13 committed:
Added delete method

JustinPeng13 committed:
Add save and storage functionality

Task data is lost between Duke sessions.

Users will want to keep their task data without running Duke 24/7.

Let's
* automatically store task data into a data.txt file whenever the task
list changes
* allow users to load tasks from a specified file on start-up, with the
default being data.txt

JustinPeng13 committed:
Enforce datetime input with time zone for Deadline and Event

Duke accepts any string as a datetime qualifier.

Standardising the datetime input will keep task descriptions clean and
allow them to be compared with each other.

Let's
* ask the user for their time zone before starting Duke
* store datetime qualifiers as ZonedDateTime instances in Deadline and
Event instances

JustinPeng13 committed:
Merge branch 'branch-Level-7'

JustinPeng13 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

JustinPeng13 committed:
Reformat indentations according to CS2103T coding style

JustinPeng13 committed:
Store commands in an enum

JustinPeng13 committed:
Abstract user-facing methods into Ui class

JustinPeng13 committed:
Create Storage and TaskList classes

JustinPeng13 committed:
Create Parser class

JustinPeng13 committed:
Organize classes into packages

JustinPeng13 committed:
Allow user to set save file on startup

JustinPeng13 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

JustinPeng13 committed:
Refactor build.gradle mainClassName

JustinPeng13 committed:
Add minimum JUnit tests

JustinPeng13 committed:
Handle reading from empty data.txt

JustinPeng13 committed:
Add JavaDoc comments for all non-private methods.

JustinPeng13 committed:
Comply to coding standard

JustinPeng13 committed:
Complete JavaDoc comments

JustinPeng13 committed:
Add find command

JustinPeng13 committed:
Merge branch 'branch-A-JavaDoc'

JustinPeng13 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java

JustinPeng13 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

JustinPeng13 committed:
Add CheckStyle and fix errors

JustinPeng13 committed:
Complete JavaDoc comments

SpecOps2016 reviewed (commented):
Very well done! Easy to read with good verb usage for function names and variables with nouns that generally make sense.

yeoyujie commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,284 @@
+            int taskNum = Integer.parseInt(fullCommand[1]) - 1;
+            Task t = tasks.get(taskNum);
+            t.markAsDone();
+            System.out.println("Nice! I've marked this task as done:\n" + t);
```
Should this be extracted out into Ui.java?

chunkhaik replied:
Speaking for all similiar cases, perhaps these outputs could be handled by a class? like a ChatBot class for better encapsulation

yeoyujie commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,284 @@
+            int taskNum = Integer.parseInt(fullCommand[1]) - 1;
+            Task t = tasks.get(taskNum);
+            t.markAsUndone();
+            System.out.println("OK, I've marked this task as not done yet:\n" + t);
```
Same as above, could this be perhaps extracted into Ui.java?

yeoyujie commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,32 @@
+     * @return 'T'
+     */
+    @Override
+    public char getType() {
```
Should a new field be created to store the type ?

yeoyujie commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,114 @@
+ */
+public class Storage {
+    /** Path to {@code .txt} save file */
+    private String saveFilePath = "data.txt";
```
should this be final and made into a constant?

yeoyujie commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,155 @@
+
+        if (scanner.hasNextLine() && scanner.nextLine().equalsIgnoreCase("Y")) {
+            while (!isValidAnswer) {
+                System.out.println("What is your timezone relative to GMT? (+/-HH:mm)");
```
I like this feature to be able to set the timezone.

yeoyujie reviewed (changes requested):
LGTM, just some minor changes to extract out certain parts of the code to the Ui.

chunkhaik commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,284 @@
+     * Starts MakiBot.
+     */
+    protected void start() {
+        System.out.println("Hello! I'm MakiBot");
```
Given the bot's called MakiBot, updating the project name would really make sense!

chunkhaik commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,284 @@
+     * @throws DukeException If the input is invalid.
+     */
+    protected void mark(String[] fullCommand) throws DukeException {
+        if (fullCommand.length < 2 || fullCommand[1].equals("")) {
```
personally find this task a bit unnecessary, either condition would be sufficient

chunkhaik commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,114 @@
+                            ZonedDateTime.parse(dataArr[3]).withZoneSameInstant(timeZone));
+                } else if (taskType == 'E') {
+                    newTask = new Event(taskDescription,
+                            ZonedDateTime.parse(dataArr[3]).withZoneSameInstant(timeZone));
+                } else if (taskType == 'T') {
+                    newTask = new Todo(taskDescription);
+                } else {
+                    System.out.println("The following task could not be loaded from memory:\n"
+                            + Arrays.toString(dataArr));
+                    continue;
+                }
+
+                if (isDone) {
+                    newTask.markAsDone();
+                }
+
+                tasks.add(newTask);
+            }
+
```
Indentation seems really off here, not sure whether it's due to github or the code, but would be good to check on it!

chunkhaik commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+     * @param keywords Words to check.
+     * @return {@code true} if task description contains all given keywords, else {@code false}.
+     */
+    private boolean containsAllKeywords(Task task, String[] keywords) {
```
great function! Would probably be adding this checker to my own project too haha. But you could probably improve it to be more specific. Like to output a string, corresponding to which keyword that's missing to improve efficiency!

chunkhaik reviewed (commented):
Overall great work! Cool new features and interesting implementation of some functions!

JustinPeng13 committed:
Implement echo user GUI

JustinPeng13 committed:
Implement full MakiBot with GUI

JustinPeng13 committed:
Merge branch 'branch-Level-10'

JustinPeng13 committed:
Clean up code and encapsulation

JustinPeng13 committed:
Fix Checkstyle

JustinPeng13 committed:
Use Varargs

JustinPeng13 committed:
Merge branch 'branch-A-Varargs'

JustinPeng13 committed:
Fix gradle main class name

JustinPeng13 committed:
Add Java assertions

The code makes certain assumptions.

Adding Java assert statements can test the correctness of any
assumptions that have been made in the program.

JustinPeng13 committed:
Improve code quality

JustinPeng13 committed:
Merge pull request #2 from JustinPeng13/branch-A-Assertions

Add Java assertions

JustinPeng13 committed:
Merge pull request #3 from JustinPeng13/branch-A-CodeQuality

Improve code quality

JustinPeng13 committed:
Add workflow to check with gradle

JustinPeng13 committed:
Add Activity subclass of Task

The Activity class allows users to specify a fixed duration in hours
that a task will take to complete.

JustinPeng13 committed:
Merge branch 'branch-BCD-Extension'

JustinPeng13 committed:
Make Parser methods static

Instances of the Parser class do not store any instance-level
information, so we do not need to store an instance in the Duke bot.

JustinPeng13 committed:
Add storage functions for Activity class

JustinPeng13 committed:
Implement auto close on "bye" command

JustinPeng13 committed:
Change bot pictures

JustinPeng13 committed:
Add UI title and format dialog box

JustinPeng13 committed:
Add Ui screenshot and User Guide

JustinPeng13 committed:
Update user guide

JustinPeng13 committed:
Fix formatting issues in user guide
